flink-streaming-java/src/test/java/org/apache/flink/streaming/api/graph/RuntimeExecutionModeDetectionTest.java

```diff
@@ -0,0 +1,158 @@
+		assertEquals(Boundedness.BOUNDED, bounded.getBoundedness());
+
+		final StreamGraph graph = createGraphAndDetectRuntimeExecutionMode(RuntimeExecutionMode.AUTOMATIC, bounded);
+		assertEquals(GlobalDataExchangeMode.POINTWISE_EDGES_PIPELINED, graph.getGlobalDataExchangeMode());
+		assertEquals(ScheduleMode.LAZY_FROM_SOURCES_WITH_BATCH_SLOT_REQUEST, graph.getScheduleMode());
+		assertFalse(graph.isAllVerticesInSameSlotSharingGroupByDefault());
+	}
+
+	@Test
+	public void testUnboundedDetection() {
+		final SourceTransformation<Integer> unbounded =
+				getSourceTransformation("Unbounded Source", Boundedness.CONTINUOUS_UNBOUNDED);
+		assertEquals(Boundedness.CONTINUOUS_UNBOUNDED, unbounded.getBoundedness());
+
+		final StreamGraph graph = createGraphAndDetectRuntimeExecutionMode(RuntimeExecutionMode.AUTOMATIC, unbounded);
+		assertEquals(GlobalDataExchangeMode.ALL_EDGES_PIPELINED, graph.getGlobalDataExchangeMode());
+		assertEquals(ScheduleMode.EAGER, graph.getScheduleMode());
+		assertTrue(graph.isAllVerticesInSameSlotSharingGroupByDefault());
+	}
+
+	@Test
+	public void testMixedDetection() {
+		final SourceTransformation<Integer> unbounded =
+				getSourceTransformation("Unbounded Source", Boundedness.CONTINUOUS_UNBOUNDED);
+		assertEquals(Boundedness.CONTINUOUS_UNBOUNDED, unbounded.getBoundedness());
+
+		final SourceTransformation<Integer> bounded =
+				getSourceTransformation("Bounded Source", Boundedness.BOUNDED);
+		assertEquals(Boundedness.BOUNDED, bounded.getBoundedness());
+
+		final StreamGraph graph = createGraphAndDetectRuntimeExecutionMode(RuntimeExecutionMode.AUTOMATIC, unbounded);
+		assertEquals(GlobalDataExchangeMode.ALL_EDGES_PIPELINED, graph.getGlobalDataExchangeMode());
+		assertEquals(ScheduleMode.EAGER, graph.getScheduleMode());
+		assertTrue(graph.isAllVerticesInSameSlotSharingGroupByDefault());
+	}
+
+	@Test
+	public void testExplicitOverridesDetectedMode() {
+		final SourceTransformation<Integer> bounded =
+				getSourceTransformation("Bounded Source", Boundedness.BOUNDED);
+		assertEquals(Boundedness.BOUNDED, bounded.getBoundedness());
+
+		final StreamGraph graph = createGraphAndDetectRuntimeExecutionMode(RuntimeExecutionMode.AUTOMATIC, bounded);
+		assertEquals(GlobalDataExchangeMode.POINTWISE_EDGES_PIPELINED, graph.getGlobalDataExchangeMode());
+		assertEquals(ScheduleMode.LAZY_FROM_SOURCES_WITH_BATCH_SLOT_REQUEST, graph.getScheduleMode());
+		assertFalse(graph.isAllVerticesInSameSlotSharingGroupByDefault());
+
+
+		final StreamGraph streamingGraph = createGraphAndDetectRuntimeExecutionMode(RuntimeExecutionMode.STREAMING, bounded);
+		assertEquals(GlobalDataExchangeMode.ALL_EDGES_PIPELINED, streamingGraph.getGlobalDataExchangeMode());
+		assertEquals(ScheduleMode.EAGER, streamingGraph.getScheduleMode());
+		assertTrue(streamingGraph.isAllVerticesInSameSlotSharingGroupByDefault());
+	}
+
+	private StreamGraph createGraphAndDetectRuntimeExecutionMode(
```

Comment:
A nit, but this method doesn't detect the runtime mode, it's just an internal detail of `generate()`. Could be just `createStreamGraph(...)`.